### ISSUE TITLE
strange behavior from IOBuffer

### ISSUE THREAD

**Issue by ExpandingMan** [JuliaLang/Compat.jl#502]
The `Compat` version of `IOBuffer` does not seem to be working properly on 0.6.  If I do
```julia
b = IOBuffer(zeros(UInt8, 4), write=true)
write(b, 'a')

take!(b)
```
On the last line on 0.7 I get simply `[0x61]` (expected behavior).  On 0.6 however I get the entire buffer: `[0x61,0x00,0x00,0x00]`.

The problem is even worse if you do `IOBuffer(maxsize=4, write=true)` in which case in 0.6 all the values of an uninitialized array are returned.

Perhaps [this line](https://github.com/JuliaLang/Compat.jl/blob/master/src/Compat.jl#L264) is calling the wrong function?

**Comment by stevengj** [comment]
I think #501 is buggy, as I mentioned at the end of that PR...

**Comment by stevengj** [comment]
I'm preparing a PR to fix IOBuffer, including this issue.

**Comment by ExpandingMan** [comment]
This was resolved by #504.
